packages/flutter/lib/src/material/bottom_navigation_bar.dart

```diff
@@ -0,0 +1,551 @@
+                      )
+                    ),
+                    new Align(
+                      alignment: FractionalOffset.bottomCenter,
+                      child: new Container(
+                        margin: const EdgeInsets.only(bottom: 10.0),
+                        child: new FadeTransition(
+                          opacity: _animations[i],
+                          child: new DefaultTextStyle(
+                            style: new TextStyle(
+                              fontSize: 14.0,
+                              color: Colors.white
+                            ),
+                            child: config.labels[i].title
+                          )
+                        )
+                      )
+                    )
+                  ]
+                )
+              )
+            )
+          );
+        }
+
+        bottomNavigation = new SizedBox(
+          width: _maxWidth,
+          child: new Row(
+            children: children
+          )
+        );
+
+        break;
+    }
+
+    return new Stack(
+      children: <Widget>[
+        new Positioned.fill(
+          child: new Material(
+            elevation: 8,
+            color: _backgroundColor
+          )
+        ),
+        new SizedBox(
+          height: kBottomNavigationBarHeight,
+          child: new Center(
+            child: new Stack(
+              children: <Widget>[
+                new Positioned(
+                  left: 0.0,
+                  top: 0.0,
+                  right: 0.0,
+                  bottom: 0.0,
+                  child: new CustomPaint(
+                    painter: new _RadialPainter(
+                      circles: _circles.toList()
+                    )
+                  )
+                ),
+                new Material(
+                  type: MaterialType.transparency,
+                  child: new Center(
+                    child: bottomNavigation
+                  )
+                )
+              ]
+            )
+          )
+        )
+      ]
+    );
+  }
+}
+
+class _Circle {
+  _Circle({
+    this.state,
+    this.index,
+    this.color
+  }) {
+    controller = new AnimationController(
+      duration: kThemeAnimationDuration
+    );
+    animation = new CurvedAnimation(
+      parent: controller,
+      curve: Curves.fastOutSlowIn
+    );
+
+    controller.forward();
+  }
+
+  final _BottomNavigationBarState state;
+  final int index;
+  final Color color;
+  AnimationController controller;
+  CurvedAnimation animation;
+
+  FractionalOffset get offset {
+    return state.cirleOffset(index);
+  }
+
+  void dispose() {
+    controller.dispose();
+  }
+
+  @override
+  int get hashCode => hashValues(offset, controller, color);
+
+  @override
+  bool operator ==(dynamic other) {
+    if (identical(this, other))
+      return true;
+    if (other is! _Circle)
+      return false;
+    final _Circle typedOther = other;
+    return offset == typedOther.offset &&
```

Comment:
Since the constructor creates the controller, can two _Circles be == if they're not identical? In other words, this override  of == doesn't seem necessary.